File: algorithms/sorting/radix-sort/RadixSort.java
public final class RadixSort {
    private RadixSort() {
    }

    public static int[] radixSort(int[] numbers) {
        if (numbers.length == 0) return new int[] {};

        int minimum = numbers[0];
        for (int value : numbers) minimum = Math.min(minimum, value);
        long offset = minimum < 0 ? -(long) minimum : 0;
        long[] array = new long[numbers.length];
        long maximum = 0;
        for (int index = 0; index < numbers.length; index++) {
            array[index] = numbers[index] + offset;
            maximum = Math.max(maximum, array[index]);
        }

        for (long place = 1; maximum / place > 0; place *= 10) {
            int[] counts = new int[10];
            for (long value : array) counts[(int) ((value / place) % 10)]++;
            for (int digit = 1; digit < 10; digit++) counts[digit] += counts[digit - 1];

            long[] output = new long[array.length];
            for (int index = array.length - 1; index >= 0; index--) {
                int digit = (int) ((array[index] / place) % 10);
                counts[digit]--;
                output[counts[digit]] = array[index];
            }
            array = output;
            if (place > maximum / 10) break;
        }

        int[] result = new int[array.length];
        for (int index = 0; index < array.length; index++) result[index] = (int) (array[index] - offset);
        return result;
    }
}

File: algorithms/sorting/radix-sort/radix-sort.mjs
export function radixSort(numbers) {
  if (numbers.some((value) => !Number.isSafeInteger(value))) {
    throw new TypeError("Radix Sort requires safe integer values");
  }
  if (numbers.length === 0) return [];

  const minimum = Math.min(...numbers);
  const offset = minimum < 0 ? -minimum : 0;
  let array = numbers.map((value) => value + offset);
  if (array.some((value) => !Number.isSafeInteger(value))) {
    throw new RangeError("Shifted keys must remain safe integers");
  }

  const maximum = Math.max(...array);
  for (let place = 1; Math.floor(maximum / place) > 0; place *= 10) {
    const counts = Array(10).fill(0);
    for (const value of array) counts[Math.floor(value / place) % 10] += 1;
    for (let digit = 1; digit < 10; digit += 1) counts[digit] += counts[digit - 1];

    const output = Array(array.length);
    for (let index = array.length - 1; index >= 0; index -= 1) {
      const digit = Math.floor(array[index] / place) % 10;
      counts[digit] -= 1;
      output[counts[digit]] = array[index];
    }
    array = output;
    if (place > Math.floor(maximum / 10)) break;
  }

  return array.map((value) => value - offset);
}
